### PR TITLE
[EXAMPLE]Add multi gpu graph predication GIN+virtualnode example

### PR DESCRIPTION
**Pull request by peizhou001**
## Description
Add a new multi node graph predication example with GIN+virtualnode model and OGB dataset.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [x] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/pytorch/gin/multi_gpu_graph_prediction.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+        # add virtual node
+        self.virtual_emb = nn.Embedding(1, n_hidden)
+        nn.init.constant_(self.virtual_emb.weight.data, 0)
+        self.virtual_layers = nn.ModuleList()
+        for _ in range(n_layers - 1):
+            self.virtual_layers.append(MLP(n_hidden))
+        self.virtual_pool = dglnn.SumPooling()
+
+    def forward(self, g, x, adj_t):
```
Comment (jermainewang) [inline examples/pytorch/gin/multi_gpu_graph_prediction.py:52]: `adj_t` -> `x_e`

File: examples/pytorch/gin/multi_gpu_graph_prediction.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+    train_dataloader = DataLoader(
+            dataset[dataset.train_idx], batch_size=32,
+            use_ddp=True, shuffle=True)
+    valid_dataloader = DataLoader(
+            dataset[dataset.val_idx], batch_size=32)
+    test_dataloader = DataLoader(
+            dataset[dataset.test_idx], batch_size=32)
+
+    durations = []
```
Comment (jermainewang) [inline examples/pytorch/gin/multi_gpu_graph_prediction.py:106]: `durations` is not used anywhere. remove

File: examples/pytorch/gin/multi_gpu_graph_prediction.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    for epoch in range(50):
+        model.train()
+        train_dataloader.set_epoch(epoch)
+        t0 = time.time()
```
Comment (jermainewang) [inline examples/pytorch/gin/multi_gpu_graph_prediction.py:110]: same for all the timing code

File: examples/pytorch/gin/multi_gpu_graph_prediction.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        nn.init.constant_(self.virtual_emb.weight.data, 0)
+        self.virtual_layers = nn.ModuleList()
+        for _ in range(n_layers - 1):
+            self.virtual_layers.append(MLP(n_hidden))
+        self.virtual_pool = dglnn.SumPooling()
+
+    def forward(self, g, x, adj_t):
+        v_emb = self.virtual_emb.weight.expand(g.batch_size, -1)
+        hn = self.node_encoder(x)
+        for layer in range(len(self.layers)):
```
Comment (jermainewang) [inline examples/pytorch/gin/multi_gpu_graph_prediction.py:55]: `layer` -> `i` since it is more like a counter

File: examples/pytorch/gin/multi_gpu_graph_prediction.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+            v_hn = dgl.broadcast_nodes(g, v_emb)
+            hn = hn + v_hn
+            he = self.edge_encoders[layer](adj_t)
+            hn = self.layers[layer](g, hn, he)
+            if layer != len(self.layers) - 1:
```
Comment (jermainewang) [inline examples/pytorch/gin/multi_gpu_graph_prediction.py:60]: L60 - L66 read a bit strange. Why dropout is needed for all the layers while all other non-linear transformation (e.g., activation and virtual nodes) are skipped for layers other than the last one? Can we simply have:
```python
if layer != len(self.layers) - 1:
    hn = F.relu(hn)
    hn = self.dropout(hn)
    v_emb_tmp = ...
    ...
```

**Comment by jermainewang** [comment]
Apart from the comments above, how is the scalability of the code w.r.t. the number of GPUs in use?

File: examples/pytorch/gin/multi_gpu_graph_prediction.py (hunk 1 above)
Comment (peizhou001) [reply examples/pytorch/gin/multi_gpu_graph_prediction.py:52]: fixed

File: examples/pytorch/gin/multi_gpu_graph_prediction.py (hunk 2 above)
Comment (peizhou001) [reply examples/pytorch/gin/multi_gpu_graph_prediction.py:106]: removed

File: examples/pytorch/gin/multi_gpu_graph_prediction.py (hunk 3 above)
Comment (peizhou001) [reply examples/pytorch/gin/multi_gpu_graph_prediction.py:110]: removed time related code

File: examples/pytorch/gin/multi_gpu_graph_prediction.py (hunk 4 above)
Comment (peizhou001) [reply examples/pytorch/gin/multi_gpu_graph_prediction.py:55]: changed

**Comment by mufeili** [comment]
Should this go to `examples/pytorch/gin` or `examples/pytorch/ogb`?

File: examples/pytorch/gin/README.md [hunk 6]
```diff
@@ -23,11 +24,25 @@ python3 train.py --dataset MUTAG
...
 -------
 | Dataset       | Result 
 | ------------- | -------
 | MUTAG         | ~89.4    
 | PTC           | ~68.5   
 | NCI1          | ~82.9    
-| PROTEINS      | ~74.1   
+| PROTEINS      | ~74.1
+
+Multi GPU training for graph predication
```
Comment (mufeili) [inline examples/pytorch/gin/README.md:36]: graph property prediction

File: examples/pytorch/gin/multi_gpu_graph_prediction.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+            nn.BatchNorm1d(2 * in_feats),
+            nn.ReLU(),
+            nn.Linear(2 * in_feats, in_feats),
+            nn.BatchNorm1d(in_feats)
+        )
+
+    def forward(self, h):
+        return self.mlp(h)
+
+class GIN(torch.nn.Module):
```
Comment (mufeili) [inline examples/pytorch/gin/multi_gpu_graph_prediction.py:30]:
```suggestion
class GIN(nn.Module):
```

File: examples/pytorch/gin/multi_gpu_graph_prediction.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    parser.add_argument('--dataset', type=str, default="ogbg-molhiv",
+                        choices=['ogbg-molhiv', 'ogbg-molpcba'],
+                        help='name of dataset (default: ogbg-molhiv)')
+    dataset_name = parser.parse_args().dataset
+    root = './data/OGB'
+
+    world_size = torch.cuda.device_count()
+    print('Let\'s use', world_size, 'GPUs!')
+    args = (world_size, dataset_name, root)
+    import torch.multiprocessing as mp
```
Comment (mufeili) [inline examples/pytorch/gin/multi_gpu_graph_prediction.py:150]: The dataset should be loaded here as in [the tutorial](https://docs.dgl.ai/tutorials/multi/1_graph_classification.html#sphx-glr-tutorials-multi-1-graph-classification-py).

File: examples/pytorch/gin/multi_gpu_graph_prediction.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+        durations.append(tt - t0)
+
+        if rank == 0:
+            val_metric = evaluate(valid_dataloader, rank, model, evaluator)[evaluator.eval_metric]
+            test_metric = evaluate(test_dataloader, rank, model, evaluator)[evaluator.eval_metric]
+
+            print(f'Epoch: {epoch:03d}, Loss: {loss:.4f}, '
+                  f'Val: {val_metric:.4f}, Test: {test_metric:.4f}')
+
+        dist.barrier()
```
Comment (mufeili) [inline examples/pytorch/gin/multi_gpu_graph_prediction.py:134]: Do we need this?

**Comment by jermainewang** [comment]
> Should this go to `examples/pytorch/gin` or `examples/pytorch/ogb`?

Depending on the purpose of this example:
- *Is this for teaching users how to write GIN?* Yes and no, but more no IMO because (1) there is already a script and (2) it is a variant of the original GIN model in the paper.
- *Is this for reproducing OGB leaderboard?* No. The performance is not carefully tuned.
- *Is this for teaching users how to write multi-GPU graph prediction?* Yes, though we picked a model based on one of the OGB baseline, GIN + virtual node.

Therefore, I think putting the example either in `examples/pytorch/gin` or `examples/pytorch/ogb` will make it hard for users to find them. I think we probably should create a folder `examples/pytorch/multigpu/` and put the example there. We may also want to put the multigpu GraphSAGE example there too. Similarly, distributed training should have its own folder.

cc @chang-l @BarclayII @TristonC for opinions.

File: examples/pytorch/gin/multi_gpu_graph_prediction.py (hunk 5 above)
Comment (peizhou001) [reply examples/pytorch/gin/multi_gpu_graph_prediction.py:60]: Yes, I think we can simplify the model to have Relu for every layer and keep v_emb excluded in the latest layer as it is used for next layer computation but it is already the last one . For relu, originally we followed the implementation [here](https://github.com/snap-stanford/ogb/blob/master/examples/graphproppred/mol/conv.py#L205), but according my test, it performs much the same, so it's ok to simplify the code:

| Dataset     |  Contain relu in last layer | Test | Validation |
| ----------- | ----------- |
| Ogbg-molhiv      | No       | 0.7964,       |0.7904|
|  Ogbg-molhiv | Yes       | 0.7922   | 0.8092 

| Ogbg-molpcba      | Yes       | 0.2288       |0.2308|
|  Ogbg-molpcba | No       | 0.2239   | 0.2283|

**Comment by peizhou001** [comment]
>

**Comment by peizhou001** [comment]
> Apart from the comments above, how is the scalability of the code w.r.t. the number of GPUs in use?

The scalability performs well when GPU number is under 4, but the time doesn't decrease and even increase sightly when adding GPU number from 4 to 8, I'm trying to understand why it happens. And this is the averaged data after dozens of runs:

### Ogbg-molhiv
|GPU number |Acceleration ratio |  |
| --- | ----------- |
| 1 | x |
| 2 | 1.7x |
| 4 | 2.5x |
| 8 | 2.3x |

### Ogbg-molhiv
|GPU number |Acceleration ratio |  |
| --- | ----------- |
| 1 | x |
| 2 | 2.2x |
| 4 |  3.5x|
| 8 | 3.3x|

File: examples/pytorch/gin/README.md (hunk 6 above)
Comment (peizhou001) [reply examples/pytorch/gin/README.md:36]: fixed

File: examples/pytorch/gin/multi_gpu_graph_prediction.py (hunk 7 above)
Comment (peizhou001) [reply examples/pytorch/gin/multi_gpu_graph_prediction.py:30]: fixed

File: examples/pytorch/gin/multi_gpu_graph_prediction.py (hunk 8 above)
Comment (peizhou001) [reply examples/pytorch/gin/multi_gpu_graph_prediction.py:150]: Remove it accidentally, re-add for fix.

File: examples/pytorch/gin/multi_gpu_graph_prediction.py (hunk 9 above)
Comment (peizhou001) [reply examples/pytorch/gin/multi_gpu_graph_prediction.py:134]: It is not necessary as the synchronization happens in loss.backward(), so remove it for simplification.

**Comment by chang-l** [comment]
>Therefore, I think putting the example either in examples/pytorch/gin or examples/pytorch/ogb will make it hard for users to find them. I think we probably should create a folder examples/pytorch/multigpu/ and put the example there. We may also want to put the multigpu GraphSAGE example there too. Similarly, distributed training should have its own folder.

I agree. Additionally, I think the `multigpu` folder needs to be under baseline example folder eventually if its for educational purposes, similar to `distributed` or other advanced folders.

File: examples/pytorch/multigpu/README.md [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+Run with following (available dataset: "ogbg-molhiv", "ogbg-molpcba")
+```bash
+python3 multi_gpu_graph_prediction.py --dataset ogbg-molhiv
+```
+
+Results:
+```
+* ogbg-molhiv: ~0.7965
+* ogbg-molpcba: ~0.2239
+```
```
Comment (jermainewang) [inline examples/pytorch/multigpu/README.md:26]: Can you add your scalability result as a table in this README?

File: examples/pytorch/multigpu/multi_gpu_graph_prediction.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.distributed as dist
+import torch.optim as optim
+import dgl
+import dgl.nn as dglnn
+from dgl.data import AsGraphPredDataset
+from dgl.dataloading import GraphDataLoader as DataLoader
```
Comment (jermainewang) [inline examples/pytorch/multigpu/multi_gpu_graph_prediction.py:9]: This is a bit strange. Just use `GraphDataLoader` is fine. No need to create alias.

**Review by jermainewang — APPROVED** [review]
LGTM

File: examples/pytorch/multigpu/multi_gpu_graph_prediction.py (hunk 11 above)
Comment (peizhou001) [reply examples/pytorch/multigpu/multi_gpu_graph_prediction.py:9]: fixed

File: examples/pytorch/multigpu/README.md (hunk 10 above)
Comment (peizhou001) [reply examples/pytorch/multigpu/README.md:26]: sure, add it for already

File: examples/pytorch/multigpu/README.md [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+------------
+
+```bash
+pip install torchmetrics
+```
+
+How to run
+-------
+
+### Graph property predication
```
Comment (mufeili) [inline examples/pytorch/multigpu/README.md:14]: prediction

File: examples/pytorch/multigpu/README.md [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+We test scalability of the code with dataset "ogbg-molhiv" in a machine of type <a href="https://aws.amazon.com/blogs/aws/now-available-ec2-instances-g4-with-nvidia-t4-tensor-core-gpus/">Amazon EC2 g4dn.metal</a>
+, which has **8 Nvidia T4 Tensor Core GPUs**.
+
+
+|GPU number |Speed Up |Batch size |Test accuracy |Average epoch Time|
+| --- | ----------- | ----------- | -----------|-----------|
+| 1 | x | 32 | 0.7765| 45.0s|
+| 2 | 3.7x |64 | 0.7761|12.1s|
+| 4 | 5.9x| 128 |  0.7854|7.6s|
+| 8 | 9.5x| 256 |  0.0.7751|4.7s|
```
Comment (mufeili) [inline examples/pytorch/multigpu/README.md:38]: Remove extra `0.`

File: examples/pytorch/multigpu/README.md (hunk 12 above)
Comment (peizhou001) [reply examples/pytorch/multigpu/README.md:14]: fixed

File: examples/pytorch/multigpu/README.md (hunk 13 above)
Comment (peizhou001) [reply examples/pytorch/multigpu/README.md:38]: fixed

File: examples/pytorch/gin/multi_gpu_graph_prediction.py [hunk 14]
```diff
@@ -118,21 +114,14 @@ def train(rank, world_size, dataset_name, root):
...
-        tt = time.time()
 
         if rank == 0:
-            print(tt - t0)
-        durations.append(tt - t0)
-
-        if rank == 0:
-            val_metric = evaluate(valid_dataloader, rank, model, evaluator)[evaluator.eval_metric]
-            test_metric = evaluate(test_dataloader, rank, model, evaluator)[evaluator.eval_metric]
+            val_metric = evaluate(valid_dataloader, rank, model.module, evaluator)[evaluator.eval_metric]
```
Comment (mufeili) [inline examples/pytorch/gin/multi_gpu_graph_prediction.py]: Is it sufficient to pass `model`?

File: examples/pytorch/gin/multi_gpu_graph_prediction.py [hunk 15]
```diff
@@ -118,21 +114,14 @@ def train(rank, world_size, dataset_name, root):
...
+            test_metric = evaluate(test_dataloader, rank, model.module, evaluator)[evaluator.eval_metric]
```
Comment (mufeili) [inline examples/pytorch/gin/multi_gpu_graph_prediction.py]: Ditto

File: examples/pytorch/gin/multi_gpu_graph_prediction.py (hunk 14 above)
Comment (peizhou001) [reply examples/pytorch/gin/multi_gpu_graph_prediction.py]: No, according to my test, use DDP model directly may cause stuck here, so I change it to literal model.
